[Docs] Refine Quick Start (#378)
InternLM/xtuner: docs/en/get_started/quickstart.md

# Quickstart

In this section, we will show you how to use XTuner to fine-tune a model to help you get started quickly.

After installing XTuner successfully, we can start fine-tuning the model. In this section, we will demonstrate how to use XTuner to apply the QLoRA algorithm to fine-tune InternLM2-Chat-7B on the Colorist dataset.

> The Colorist dataset is a dataset that provides color choices and suggestions based on color descriptions. A model fine-tuned on this dataset can be used to give a hexadecimal color code based on the user's description of the color. For example, when the user enters "a calming but fairly bright light sky blue, between sky blue and baby blue, with a hint of fluorescence due to its brightness", the model will output ![#66ccff](https://img.shields.io/badge/%2366ccff-66CCFF), which matches the user's description. There are a few sample data from this dataset:

Reviewer: Please add dataset URL here.

| Light Sky Blue: A calming, fairly bright color that falls between sky blue and baby blue, with a hint of slight fluorescence due to its brightness.                                                                              | 浅天蓝色：一种介于天蓝和婴儿蓝之间的平和、相当明亮的颜色，由于明亮而带有一丝轻微的荧光。                                         | #66ccff: ![#66ccff](https://img.shields.io/badge/%2366ccff-66CCFF) |
| Bright red: This is a very vibrant, saturated and vivid shade of red, resembling the color of ripe apples or fresh blood. It is as red as you can get on a standard RGB color palette, with no elements of either blue or green. | 鲜红色： 这是一种非常鲜艳、饱和、生动的红色，类似成熟苹果或新鲜血液的颜色。它是标准 RGB 调色板上的红色，不含任何蓝色或绿色元素。 | #ee0000: ![#ee0000](https://img.shields.io/badge/%23ee0000-EE0000) |
| Bright Turquoise: This color mixes the freshness of bright green with the tranquility of light blue, leading to a vibrant shade of turquoise. It is reminiscent of tropical waters.                                              | 明亮的绿松石色：这种颜色融合了鲜绿色的清新和淡蓝色的宁静，呈现出一种充满活力的绿松石色调。它让人联想到热带水域。                 | #00ffcc: ![#00ffcc](https://img.shields.io/badge/%2300ffcc-00FFCC) |

## Prepare the model weights

> Before fine-tuning the model, we first need to prepare the weights of the model. Since XTuner pulls model weights from HuggingFace by default, which may lead to an unstable download process, slow download speed and other problems, we choose to download the weights of InternLM2-Chat-7B from ModelScope.

Reviewer: Please give two download examples. The first is from HuggingFace, and the second is from ModelScope.
Specify that only users experiencing network issues should consider using ModelScope.

```bash
pip install modelscope

# Download the model weights to the current directory
python -c "from modelscope import snapshot_download; snapshot_download('Shanghai_AI_Laboratory/internlm2-chat-7b', cache_dir='.')"
```

After completing the download, we can start to prepare the dataset for fine-tuning.

The HuggingFace link and ModelScope link are attached here:

- The HuggingFace link is located at: https://huggingface.co/internlm/internlm2-chat-7b
- The ModelScope link is located at: https://modelscope.cn/models/Shanghai_AI_Laboratory/internlm2-chat-7b/summary

## Prepare the fine-tuning dataset

Due to the same reason, we choose to download the dataset from ModelScope.

```bash
git clone https://www.modelscope.cn/datasets/fanqiNO1/colors.git

> ```

Reviewer: Please give two examples, one is from HuggingFace and another one is from ModelScope.

#                          PART 1  Settings                           #
#######################################################################
# Model
# pretrained_model_name_or_path = 'internlm/internlm2-7b'
pretrained_model_name_or_path = './Shanghai_AI_Laboratory/internlm2-chat-7b'

# Data
# data_path = 'burkelibbey/colors'
data_path = './colors/train.jsonl'
# prompt_template = PROMPT_TEMPLATE.default
prompt_template = PROMPT_TEMPLATE.internlm2_chat

...
#######################################################################
#                      PART 3  Dataset & Dataloader                   #
#######################################################################
train_dataset = dict(
    type=process_hf_dataset,
    # dataset=dict(type=load_dataset, path=data_path),
    dataset=dict(type=load_dataset, path='json', data_files=dict(train=data_path)),
    tokenizer=tokenizer,
    max_length=max_length,
    dataset_map_fn=colors_map_fn,
    template_map_fn=dict(
        type=template_map_fn_factory, template=prompt_template),
    remove_unused_columns=True,
    shuffle_before_pack=True,
    pack_to_max_length=pack_to_max_length)

> ```

Reviewer: Please use `diff` markdown block (like [here](https://github.com/InternLM/xtuner/tree/main/examples/demo_data/multi_turn_2#config)), instead of `python + comment`

Therefore, `pretrained_model_name_or_path`, `data_path`, `prompt_template`, and the `dataset` fields in `train_dataset` are modified.

## Start fine-tuning

Once having done the above steps, we can start fine-tuning using the following command.

> ```bash
> # Single GPU
> xtuner train internlm2_7b_qlora_colorist_e5_copy.py
> # Multiple GPUs
> NPROC_PER_NODE=${GPU_NUM} xtuner train internlm2_7b_qlora_colorist_e5_copy.py
> # Slurm
> srun ${SRUN_ARGS} xtuner train internlm2_7b_qlora_colorist_e5_copy.py --launcher slurm
> ```

Reviewer: ```suggestion
```bash
# Single GPU
xtuner train ./internlm2_7b_qlora_colorist_e5_copy.py
# Multiple GPUs
NPROC_PER_NODE=${GPU_NUM} xtuner train ./internlm2_7b_qlora_colorist_e5_copy.py
# Slurm
srun ${SRUN_ARGS} xtuner train ./internlm2_7b_qlora_colorist_e5_copy.py --launcher slurm
```
```

```

After training, the output of the model is as shown below:

```text
2024/01/29 22:08:07 - mmengine - INFO - Sample output:
<s><|im_start|>system
You are a professional color designer. Please provide the corresponding colors based on the description of Human.
<|im_end|>
<|im_start|>user
请给我一个像天空一样清澈透明的蓝色。<|im_end|>
<|im_start|>assistant
#1099ee<|im_end|>

2024/01/29 22:08:08 - mmengine - INFO - Sample output:
<s><|im_start|>system
You are a professional color designer. Please provide the corresponding colors based on the description of Human.
<|im_end|>
<|im_start|>user
Please give me a clear blue like the sky.<|im_end|>
<|im_start|>assistant
#0066dd<|im_end|>
```

The color of the model output is shown below:

- 天空一样清澈透明的蓝色：![天空一样清澈透明的蓝色](https://img.shields.io/badge/天空一样清澈透明的蓝色-1099EE)
- A clear blue like the sky: ![A clear blue like the sky](https://img.shields.io/badge/A_clear_blue_like_the_sky-0066DD)

It is clear that the output of the model after training has been fully aligned with the content of the dataset.

# Model Convert + LoRA Merge

> After training, we will get several pth files that do **not** contain all the parameters of the model, but store the parameters updated by the training process of the QLoRA algorithm. Therefore, we need to convert these pth files to hf format and merge them into the original weights.

Reviewer: ```suggestion
After training, we will get several `.pth` files that do **NOT** contain all the parameters of the model, but store the parameters updated by the training process of the QLoRA algorithm. Therefore, we need to convert these `.pth` files to HuggingFace format and merge them into the original LLM weights.
```

### Model Convert

> XTuner has already integrated the function of converting the model to hf format. We can use the following command to convert the model.

Reviewer: ```suggestion
XTuner has already integrated the tool of converting the model to HuggingFace format. We can use the following command to convert the model.
```

```bash
# Create the directory to store parameters in hf format
mkdir work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf

# Convert the model to hf format

> xtuner convert pth_to_hf internlm2_7b_qlora_colorist_e5_copy.py\
>                             work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720.pth\
>                             work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf

Reviewer: ```suggestion
xtuner convert pth_to_hf internlm2_7b_qlora_colorist_e5_copy.py \
                            work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720.pth \
                            work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf
```

```

> This command will convert `work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720.pth` to hf format based on the contents of the config `internlm2_7b_qlora_colorist_e5_copy/iter_720.pth` and will save it in `work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf`.

Reviewer: ```suggestion
This command will convert `work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720.pth` to hf format based on the contents of the config `internlm2_7b_qlora_colorist_e5_copy.py` and will save it in `work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf`.
```

### LoRA Merge

> XTuner has also integrated the function of merging weights, we just need to execute the following command:

Reviewer: ```suggestion
XTuner has also integrated the tool of merging LoRA weights, we just need to execute the following command:
```

```bash
# Create the directory to store the merged weights
mkdir work_dirs/internlm2_7b_qlora_colorist_e5_copy/merged

# Merge the weights

> xtuner convert merge Shanghai_AI_Laboratory/internlm2-chat-7b\
>                         work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf\
>                         work_dirs/internlm2_7b_qlora_colorist_e5_copy/merged\
>                         --max-shard-size 2GB

Reviewer: ```suggestion
xtuner convert merge Shanghai_AI_Laboratory/internlm2-chat-7b \
                        work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf \
                        work_dirs/internlm2_7b_qlora_colorist_e5_copy/merged \
                        --max-shard-size 2GB
```

```

Similar to the command above, this command will read the original parameter path `Shanghai_AI_Laboratory/internlm2-chat-7b` and the path of parameter which has been converted to hf format `work_dirs/internlm2_7b_qlora_colorist_e5_copy/iter_720_hf` and merge the two parts of the parameters and save them in `work_dirs/internlm2_7b_qlora_colorist_e5_copy/merged`, where the maximum file size for each parameter slice is 2GB.

## Chat with the model

> To better appreciate the model's capabilities after merging the weights, we can chat with the model. XTuner also integrates the function of chatting with the merged model. We can start a simple demo to chat with the model with the following command:

Reviewer: ```suggestion
To better appreciate the model's capabilities after merging the weights, we can chat with the model. XTuner also integrates the tool of chatting with models. We can start a simple demo to chat with the model with the following command:
```

```bash

> xtuner chat work_dirs/internlm2_7b_qlora_colorist_e5_copy/merged\
>                 --prompt-template internlm2_chat\
>                 --system-template colorist

Reviewer: ```suggestion
xtuner chat work_dirs/internlm2_7b_qlora_colorist_e5_copy/merged \
                --prompt-template internlm2_chat \
                --system-template colorist
```

```

> 

Reviewer: Add an example of chatting with a LLM and LoRA adapter here will be better!
